index.js

```diff
@@ -192,6 +193,93 @@ module.exports = class HyperDHT extends DHT {
+
+  async immutablePut (value) {
+    if (Buffer.isBuffer(value) === false) throw new Error('value must be a buffer')
+    if (value.length > PUT_VALUE_MAX_SIZE) {
+      throw new Error(`Value size must be <= ${PUT_VALUE_MAX_SIZE}`)
+    }
+    const key = Buffer.allocUnsafe(32)
+    sodium.crypto_generichash(key, value)
+    const query = this.query(key, 'immutable_get', null, {
+      map: mapImmutable,
+      commit (node, dht) {
+        return dht.request(key, 'immutable_put', value, node.from, { map: mapImmutable, token: node.token })
+      }
+    })
+    await query.finished()
+    return { key }
+  }
+
+  async mutableGet (key, { salt = null, seq = 0, latest = true } = {}) {
+    if (Buffer.isBuffer(key) === false) throw new Error('key must be a buffer')
+    if (typeof seq !== 'number') throw new Error('seq should be a number')
+    if (salt) {
+      if (Buffer.isBuffer(salt) === false) throw new Error('salt must be a buffer')
+      if (salt.length > 64) throw new Error('salt size must be no greater than 64 bytes')
+    }
+
+    const request = cenc.encode(messages.mutableReq, { salt, seq })
+
+    const query = this.query(key, 'mutable_get', request, { map: mapMutable })
+    const userSeq = seq
+    let topSeq = seq
+    let result = null
+    for await (const node of query) {
+      const { value: item, id, ...meta } = node
+      const { value, signature, seq: storedSeq, publicKey } = cenc.decode(messages.mutableRes, item)
+      const msg = hypersign.signable(value, { salt, seq: storedSeq })
+      if (storedSeq >= userSeq && sodium.crypto_sign_verify_detached(signature, msg, publicKey)) {
+        if (latest === false) return { id, value, signature, seq: storedSeq, salt, ...meta }
+        if (storedSeq >= topSeq) {
+          topSeq = storedSeq
+          result = { id, value, signature, seq: storedSeq, salt, ...meta }
+        }
+      }
+    }
+    return result
+  }
+
+  async mutablePut (value, opts = {}) {
+    if (Buffer.isBuffer(value) === false) throw new Error('value must be a buffer')
+    if (value.length > PUT_VALUE_MAX_SIZE) {
+      throw new Error(`Value size must be <= ${PUT_VALUE_MAX_SIZE}`)
+    }
+    const { seq = 0, salt = null, keyPair, signature = hypersign.sign(value, { ...opts, keypair: keyPair }) } = opts
+    if (typeof seq !== 'number') throw new Error('seq should be a number')
+    if (opts.signature) {
+      if (!keyPair) throw new Error('keyPair is required')
+      const { secretKey, publicKey } = keyPair
+      if (Buffer.isBuffer(publicKey) === false) throw new Error('keyPair.publicKey is required')
+      if (secretKey) throw new Error('only opts.signature OR opts.keyPair.secretKey should be supplied')
+    }
+    const { publicKey } = keyPair
+    const key = Buffer.allocUnsafe(32)
+    sodium.crypto_generichash(key, publicKey)
+    const msg = cenc.encode(messages.mutableRes, {
+      value, signature, seq, salt, publicKey
+    })
+    const query = this.query(key, 'mutable_get', null, {
```

Comment:
forward opts.closestNodes here

Reply:
can I just spread the opts as in the other case